File: src/org/genyris/interp/builtin/BuiltinFunction.java
package org.genyris.interp.builtin;

import org.genyris.core.Exp;
import org.genyris.exception.GenyrisException;
import org.genyris.interp.ApplicableFunction;
import org.genyris.interp.Closure;
import org.genyris.interp.Environment;
import org.genyris.interp.Interpreter;
import org.genyris.interp.UnboundException;
import org.genyris.load.LoadFunction;

public abstract class BuiltinFunction  extends ApplicableFunction {

    public BuiltinFunction(Interpreter interp, String name, boolean eager) {
        super(interp, name, eager);
    }

    public abstract Exp bindAndExecute(Closure proc, Exp[] arguments, Environment envForBindOperations) throws GenyrisException;

    public static void bindFunctionsAndMethods(Interpreter interpreter) throws UnboundException, GenyrisException {
        interpreter.bindGlobalProcedureInstance(new ApplyFunction(interpreter));
        interpreter.bindGlobalProcedureInstance(new AsStringFunction(interpreter));
        interpreter.bindGlobalProcedureInstance(new BackquoteFunction(interpreter));
        interpreter.bindGlobalProcedureInstance(new BoundFunction(interpreter));
        interpreter.bindGlobalProcedureInstance(new CarFunction(interpreter));
        interpreter.bindGlobalProcedureInstance(new CatchFunction(interpreter));
        interpreter.bindGlobalProcedureInstance(new CdrFunction(interpreter));
        interpreter.bindGlobalProcedureInstance(new ConditionalFunction(interpreter));
        interpreter.bindGlobalProcedureInstance(new ConsFunction(interpreter));
        interpreter.bindGlobalProcedureInstance(new DefFunction(interpreter));
        interpreter.bindGlobalProcedureInstance(new DefineClassFunction(interpreter));
        interpreter.bindGlobalProcedureInstance(new DefineFunction(interpreter));
        interpreter.bindGlobalProcedureInstance(new DefMacroFunction(interpreter));
        interpreter.bindGlobalProcedureInstance(new DynamicSymbolValueFunction(interpreter));
        interpreter.bindGlobalProcedureInstance(new EqFunction(interpreter));
        interpreter.bindGlobalProcedureInstance(new EqualsFunction(interpreter));
        interpreter.bindGlobalProcedureInstance(new EvalFunction(interpreter));
        interpreter.bindGlobalProcedureInstance(new GensymFunction(interpreter));
        interpreter.bindGlobalProcedureInstance(new IdentityFunction(interpreter));
        interpreter.bindGlobalProcedureInstance(new IsFunction(interpreter));
        interpreter.bindGlobalProcedureInstance(new InternFunction(interpreter));
        interpreter.bindGlobalProcedureInstance(new LambdaFunction(interpreter));
        interpreter.bindGlobalProcedureInstance(new LambdamFunction(interpreter));
        interpreter.bindGlobalProcedureInstance(new LambdaqFunction(interpreter));
        interpreter.bindGlobalProcedureInstance(new LengthFunction(interpreter));
        interpreter.bindGlobalProcedureInstance(new ListFunction(interpreter));
        interpreter.bindGlobalProcedureInstance(new LoadFunction(interpreter));
        interpreter.bindGlobalProcedureInstance(new NthFunction(interpreter));
        interpreter.bindGlobalProcedureInstance(new ObjectFunction(interpreter));
        interpreter.bindGlobalProcedureInstance(new QuoteFunction(interpreter));
        interpreter.bindGlobalProcedureInstance(new RaiseFunction(interpreter));
        interpreter.bindGlobalProcedureInstance(new RemoveTagFunction(interpreter));
        interpreter.bindGlobalProcedureInstance(new ReplaceCarFunction(interpreter));
        interpreter.bindGlobalProcedureInstance(new ReplaceCdrFunction(interpreter));
        interpreter.bindGlobalProcedureInstance(new ReverseFunction(interpreter));
        interpreter.bindGlobalProcedureInstance(new SetFunction(interpreter));
        interpreter.bindGlobalProcedureInstance(new SymbolValueFunction(interpreter));
        interpreter.bindGlobalProcedureInstance(new SymListFunction(interpreter));
        interpreter.bindGlobalProcedureInstance(new TagFunction(interpreter));
        interpreter.bindGlobalProcedureInstance(new WhileFunction(interpreter));
        interpreter.bindGlobalProcedureInstance(new BackTraceFunction(interpreter));
        interpreter.bindGlobalProcedureInstance(new UseFunction(interpreter));
    }

}

File: src/org/genyris/interp/builtin/BuiltinFunction.kt
package org.genyris.interp.builtin

import org.genyris.core.Exp
import org.genyris.exception.GenyrisException
import org.genyris.interp.*
import org.genyris.load.LoadFunction

abstract class BuiltinFunction(interp: Interpreter, name: String?, eager: Boolean) :
    ApplicableFunction(interp, name, eager) {
    @kotlin.Throws(GenyrisException::class)
    abstract override fun bindAndExecute(
        proc: Closure?,
        arguments: Array<Exp?>?,
        envForBindOperations: Environment?
    ): Exp?

    companion object {
        @kotlin.Throws(UnboundException::class, GenyrisException::class)
        fun bindFunctionsAndMethods(interpreter: Interpreter) {
            interpreter.bindGlobalProcedureInstance(ApplyFunction(interpreter))
            interpreter.bindGlobalProcedureInstance(TailCallFunction(interpreter))
            interpreter.bindGlobalProcedureInstance(AsStringFunction(interpreter))
            interpreter.bindGlobalProcedureInstance(BackquoteFunction(interpreter))
            interpreter.bindGlobalProcedureInstance(BoundFunction(interpreter))
            interpreter.bindGlobalProcedureInstance(CarFunction(interpreter))
            interpreter.bindGlobalProcedureInstance(CatchFunction(interpreter))
            interpreter.bindGlobalProcedureInstance(CdrFunction(interpreter))
            interpreter.bindGlobalProcedureInstance(ConditionalFunction(interpreter))
            interpreter.bindGlobalProcedureInstance(ConsFunction(interpreter))
            interpreter.bindGlobalProcedureInstance(DefFunction(interpreter))
            interpreter.bindGlobalProcedureInstance(DefineClassFunction(interpreter))
            interpreter.bindGlobalProcedureInstance(DefineFunction(interpreter))
            interpreter.bindGlobalProcedureInstance(DefMacroFunction(interpreter))
            interpreter.bindGlobalProcedureInstance(DynamicSymbolValueFunction(interpreter))
            interpreter.bindGlobalProcedureInstance(EqFunction(interpreter))
            interpreter.bindGlobalProcedureInstance(EqualsFunction(interpreter))
            interpreter.bindGlobalProcedureInstance(EvalFunction(interpreter))
            interpreter.bindGlobalProcedureInstance(GensymFunction(interpreter))
            interpreter.bindGlobalProcedureInstance(HashCodeFunction(interpreter))
            interpreter.bindGlobalProcedureInstance(IdentityFunction(interpreter))
            interpreter.bindGlobalProcedureInstance(IsFunction(interpreter))
            interpreter.bindGlobalProcedureInstance(InternFunction(interpreter))
            interpreter.bindGlobalProcedureInstance(LambdaFunction(interpreter))
            interpreter.bindGlobalProcedureInstance(LambdamFunction(interpreter))
            interpreter.bindGlobalProcedureInstance(LambdaqFunction(interpreter))
            interpreter.bindGlobalProcedureInstance(LengthFunction(interpreter))
            interpreter.bindGlobalProcedureInstance(ListFunction(interpreter))
            interpreter.bindGlobalProcedureInstance(LoadFunction(interpreter))
            interpreter.bindGlobalProcedureInstance(NthFunction(interpreter))
            interpreter.bindGlobalProcedureInstance(ObjectFunction(interpreter))
            interpreter.bindGlobalProcedureInstance(QuoteFunction(interpreter))
            interpreter.bindGlobalProcedureInstance(RaiseFunction(interpreter))
            interpreter.bindGlobalProcedureInstance(RemoveTagFunction(interpreter))
            interpreter.bindGlobalProcedureInstance(ReplaceCarFunction(interpreter))
            interpreter.bindGlobalProcedureInstance(ReplaceCdrFunction(interpreter))
            interpreter.bindGlobalProcedureInstance(ReverseFunction(interpreter))
            interpreter.bindGlobalProcedureInstance(SetFunction(interpreter))
            interpreter.bindGlobalProcedureInstance(PlingFunction(interpreter))
            interpreter.bindGlobalProcedureInstance(SortFunction(interpreter))
            interpreter.bindGlobalProcedureInstance(SymbolValueFunction(interpreter))
            interpreter.bindGlobalProcedureInstance(SymbolPrefixFunction(interpreter))
            interpreter.bindGlobalProcedureInstance(SymListFunction(interpreter))
            interpreter.bindGlobalProcedureInstance(TagFunction(interpreter))
            interpreter.bindGlobalProcedureInstance(WhileFunction(interpreter))
            interpreter.bindGlobalProcedureInstance(BackTraceFunction(interpreter))
            interpreter.bindGlobalProcedureInstance(UseFunction(interpreter))
        }
    }
}
